src/hpr_packet_reporter.erl

```diff
@@ -0,0 +1,331 @@
+-module(hpr_packet_reporter).
+
+-behaviour(gen_server).
+
+-include("./grpc/autogen/server/packet_router_pb.hrl").
+
+-export([
+    start_link/1,
+    init_ets/0,
+    report_packet/2,
+    write_packets/0,
+    upload_packets/1,
+    restart_report_interval/1,
+    stop_report_interval/0
+]).
+-export([
+    init/1,
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2,
+    terminate/2,
+    code_change/3
+]).
+
+-define(SERVER, ?MODULE).
+-define(ETS, hpr_packet_report_ets).
+
+%% 5 minute interval
+-define(DEFAULT_REPORT_INTERVAL, 300000).
+-define(MAX_FILE_SIZE, 50_000_000).
+
+-type interval_duration_ms() :: non_neg_integer().
+
+-record(state, {
+    aws_client :: aws_client:aws_client(),
+    file_path :: string(),
+    report_interval :: timer:tref() | undefined
+}).
+
+%%%===================================================================
+%%% API Functions
+%%%===================================================================
+
+start_link(Args) ->
+    gen_server:start_link({local, ?SERVER}, ?SERVER, Args, []).
+
+-spec init_ets() -> ok.
+init_ets() ->
+    ?ETS = ets:new(?ETS, [named_table, bag, public, {write_concurrency, true}]),
+    ok.
+
+-spec report_packet(hpr_packet_up:packet(), hpr_route:route()) -> ok.
+report_packet(Packet, Route) ->
+    EncodedPacket = encode_packet(Packet, Route),
+    Timestamp = erlang:system_time(millisecond),
```

Comment:
@hwahby 
It looks like you could get the same Timestamp from different calls to _erlang:system_time_.
That would make this Timestamp a bad choice for a key.

Reply:
Oops, sorry. I misread the docs for 'bag'.
Comment withdrawn!